modules/store/src/reducer_creator.ts

```diff
@@ -0,0 +1,59 @@
+import { ActionCreator, ActionType, Action } from './models';
+
+export type On<S> = (state: S, action: Action) => S;
+export type OnReducer<S, C extends ActionCreator[]> = (
+  state: S,
+  action: ActionType<C[number]>
+) => S;
+export type Reducer<S> = (state: S | undefined, action: Action) => S;
+
+export function on<C1 extends ActionCreator, S>(
+  creator1: C1,
+  reducer: OnReducer<S, [C1]>
+): { reducer: On<S>; types: string[] };
+export function on<C1 extends ActionCreator, C2 extends ActionCreator, S>(
+  creator1: C1,
+  creator2: C2,
+  reducer: OnReducer<S, [C1, C2]>
+): { reducer: On<S>; types: string[] };
+export function on<
+  C1 extends ActionCreator,
+  C2 extends ActionCreator,
+  C3 extends ActionCreator,
+  S
+>(
+  creator1: C1,
+  creator2: C2,
+  creator3: C3,
+  reducer: OnReducer<S, [C1, C2, C3]>
+): { reducer: On<S>; types: string[] };
+export function on<S>(
+  creator: ActionCreator,
+  ...rest: (ActionCreator | OnReducer<S, [ActionCreator]>)[]
+): { reducer: On<S>; types: string[] };
+export function on(
+  ...args: (ActionCreator | Function)[]
+): { reducer: Function; types: string[] } {
+  const reducer = args.pop() as Function;
+  const types = args.reduce(
+    (result, creator) => [...result, (creator as ActionCreator).type],
+    [] as string[]
+  );
+  return { reducer, types };
+}
+
+export function createReducer<S>(
+  ons: { reducer: On<S>; types: string[] }[],
+  initialState: S
+): Reducer<S> {
+  const map = new Map<string, On<S>>();
+  for (let on of ons) {
+    for (let type of on.types) {
+      map.set(type, on.reducer);
```

Comment:
Not for now, but should we add some tooling here to notify if you would add the same action type twice? Because one could copy paste an action and forget to rename the type.

Reply:
Very good point!

Reply:
opened https://github.com/ngrx/platform/issues/1758